sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -56,36 +51,167 @@ public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePol
+            }
+
+            public async ValueTask<string> GetHeaderValueAsync(HttpMessage message, bool async)
             {
-                AccessToken token = async ?
-                        await _credential.GetTokenAsync(new TokenRequestContext(_scopes, message.Request.ClientRequestId), message.CancellationToken).ConfigureAwait(false) :
-                        _credential.GetToken(new TokenRequestContext(_scopes, message.Request.ClientRequestId), message.CancellationToken);
+                bool getTokenFromCredential;
+                TaskCompletionSource<HeaderValueInfo> headerValueTcs;
+                TaskCompletionSource<HeaderValueInfo>? backgroundUpdateTcs;
+                (headerValueTcs, backgroundUpdateTcs, getTokenFromCredential) = GetTaskCompletionSources();
+
+                if (getTokenFromCredential)
+                {
+                    if (backgroundUpdateTcs != null)
+                    {
+                        HeaderValueInfo info = headerValueTcs.Task.EnsureCompleted();
+                        _ = Task.Run(() => GetHeaderValueFromCredentialInBackgroundAsync(backgroundUpdateTcs, info, message, async));
+                        return info.HeaderValue;
+                    }
+
+                    try
+                    {
+                        HeaderValueInfo info = await GetHeaderValueFromCredentialAsync(message, async);
+                        headerValueTcs.SetResult(info);
+                    }
+                    catch (Exception exception)
+                    {
+                        headerValueTcs.SetException(exception);
+                        throw;
+                    }
+                }
+
+                var headerValueTask = headerValueTcs.Task;
+                if (!headerValueTask.IsCompleted)
+                {
+                    if (async)
+                    {
+                        await headerValueTask.AwaitWithCancellation(message.CancellationToken);
+                    }
+                    else
+                    {
+                        try
+                        {
+                            headerValueTask.Wait(message.CancellationToken);
+                        }
+                        catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+                    }
+                }
 
-                _headerValue = "Bearer " + token.Token;
-                _refreshOn = token.ExpiresOn - TimeSpan.FromMinutes(2);
+                return headerValueTcs.Task.EnsureCompleted().HeaderValue;
             }
 
-            if (_headerValue != null)
+            private (TaskCompletionSource<HeaderValueInfo> tcs, TaskCompletionSource<HeaderValueInfo>? backgroundUpdateTcs, bool getTokenFromCredential) GetTaskCompletionSources()
             {
-                message.Request.SetHeader(HttpHeader.Names.Authorization, _headerValue);
+                lock (_syncObj)
+                {
+                    if (_infoTcs == null)
```

Comment:
Can you please add a few comments about what conditions in this method mean logically?

Reply:
Done